### PR TITLE
Feature/2 message retention

### PR DESCRIPTION
**Pull request by adenh93**

### CODE REVIEW

File: protocol/src/frame.rs [hunk 1]
```diff
@@ -93,13 +95,38 @@ impl Frame {
...
 
+    pub fn unwrap_retention_policy(&self) -> u64 {
+        match self {
+            Self::RegisterPublisher(payload) => payload.retention_policy,
+            Self::RegisterSubscriber(payload) => payload.retention_policy,
+            _ => panic!("Attempted to unwrap non-register Frame variant"),
+        }
+    }
+
+    pub fn unwrap_batch_size(&self) -> u32 {
```
Comment (petehayes102) [inline protocol/src/frame.rs:114]: I get that these don't have to take ownership, but maybe they should be called something else, or take ownership for symmetry?

File: protocol/src/frame.rs [hunk 2]
```diff
@@ -93,13 +95,38 @@ impl Frame {
...
-            Frame::BatchMessage(bytes) => dst.extend_from_slice(&bytes),
+            Frame::BatchMessage(payload) => bincode::serialize_into(dst.writer(), &payload)
+                .map_err(ProtocolError::SerdeError)?,
             Frame::Ok => (),
         }
 
         Ok(())
     }
 
+    pub fn unwrap_retention_policy(&self) -> u64 {
```
Comment (petehayes102) [inline protocol/src/frame.rs:106]: I get that these don't have to take ownership, but maybe they should be called something else, or take ownership for symmetry?

File: tests/tests/streams/pub_sub.rs
Comment (petehayes102) [inline tests/tests/streams/pub_sub.rs:1]: These should be reworked before we merge. AFAIK your changes to the topic `Future`s have no automated test coverage without these?

**Review by petehayes102 — CHANGES_REQUESTED** [review]
Looks awesome mate! Just some petty stuff. Great job!!

File: client/src/constants.rs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 /// The default `keep_alive` interval for a client connection.
 pub const KEEP_ALIVE_DEFAULT: u64 = 5_000;
 /// The default `retention_policy` setting for messages.
-pub const RETENTION_POLICY_DEFAULT: u64 = 0;
+pub const RETENTION_POLICY_DEFAULT: u64 = 1000 * 60 * 60 * 24;
```
[the remark this answers is not in the document]
Comment (adenh93) [reply client/src/constants.rs:6]: Resolving this conversation, as we will be implementing support for streaming with and without retention in the near future.

File: log/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+edition.workspace = true
+authors.workspace = true
+license.workspace = true
+homepage.workspace = true
+repository.workspace = true
+readme.workspace = true
+keywords.workspace = true
+categories.workspace = true
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
[the remark this answers is not in the document]
Comment (adenh93) [reply log/Cargo.toml]: Whoops! 🤣

File: log/src/message/mod.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+pub const LEN_MARKER_SIZE: usize = size_of::<u64>();
+pub const CRC_SIZE: usize = size_of::<u32>();
+pub const HEADERS_SIZE: usize =
+    size_of::<u64>() + size_of::<u32>() + size_of::<u32>() + size_of::<u64>();
+
+#[derive(Debug, Clone)]
+pub struct Message {
+    headers: Headers,
+    records: Bytes,
+    _crc: u32,
```
[the remark this answers is not in the document]
Comment (adenh93) [reply log/src/message/mod.rs:21]: I've added a TODO for now. The CRC will be implemented once we have replication. For now, it's generated for the messages in preparation for the feature.

File: tests/tests/streams/pub_sub.rs
Comment (adenh93) [reply tests/tests/streams/pub_sub.rs:1]: Fixed the tests now! 👍

**Review by petehayes102 — APPROVED** [review]

**Review by petehayes102 — APPROVED** [review]

**Pull request merged**